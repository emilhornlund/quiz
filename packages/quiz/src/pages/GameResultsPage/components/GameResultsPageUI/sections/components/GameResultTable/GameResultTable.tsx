import { IconDefinition } from '@fortawesome/fontawesome-common-types'
import { faChevronDown } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import React, { FC, useState } from 'react'

import {
  CircularProgressBar,
  CircularProgressBarKind,
  CircularProgressBarSize,
} from '../../../../../../../components'
import { classNames } from '../../../../../../../utils/helpers.ts'
import styles from '../../../GameResultsPageUI.module.scss'

export interface TableItem {
  badge: number
  value: string
  progress: number
  details?: {
    title: string
    value: string | number
    icon?: IconDefinition
    iconColor?: string
  }[]
}

export interface GameResultTableProps {
  items: TableItem[]
}

const TableRow: FC<TableItem> = ({ badge, value, progress, details }) => {
  const [showDetails, setShowDetails] = useState<boolean>(false)

  return (
    <div
      key={`${badge}`}
      className={styles.tableRow}
      onClick={(event) => {
        event.preventDefault()
        setShowDetails(!showDetails)
      }}>
      <div className={styles.main}>
        <div className={styles.badge}>{`${badge}`}</div>
        <div className={styles.value} title={value}>
          {value}
        </div>
        <div>
          <CircularProgressBar
            progress={progress}
            kind={CircularProgressBarKind.Correct}
            size={CircularProgressBarSize.Small}
          />
        </div>
        <div>
          <FontAwesomeIcon icon={faChevronDown} color="black" />
        </div>
      </div>
      <div
        className={classNames(
          styles.details,
          showDetails ? styles.active : undefined,
        )}>
        {details?.map((item) => (
          <div key={item.title} className={styles.item}>
            <div className={styles.title}>
              {item.icon && (
                <FontAwesomeIcon icon={item.icon} color={item.iconColor} />
              )}
              {item.title}
            </div>
            <div className={styles.value}>{item.value}</div>
          </div>
        ))}
      </div>
    </div>
  )
}

const GameResultTable: FC<GameResultTableProps> = ({ items }) => {
  return (
    <div className={styles.table}>
      {items.map((metric, index) => (
        <TableRow key={index} {...metric} />
      ))}
    </div>
  )
}

export default GameResultTable
