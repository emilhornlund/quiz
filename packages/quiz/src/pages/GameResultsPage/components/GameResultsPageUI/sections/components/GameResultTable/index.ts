export type {
  TableItem,
  TableSeparator,
  GameResultTableProps,
} from './GameResultTable'
export { default } from './GameResultTable'
