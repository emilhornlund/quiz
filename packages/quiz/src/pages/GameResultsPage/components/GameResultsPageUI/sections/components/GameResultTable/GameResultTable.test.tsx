import { faChevronDown } from '@fortawesome/free-solid-svg-icons'
import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { describe, expect, it } from 'vitest'

import GameResultTable, { TableItem } from './GameResultTable'

describe('GameResultTable', () => {
  const items: TableItem[] = [
    {
      type: 'table-row',
      badge: 1,
      value: 'Alice Anderson',
      progress: 82,
      details: [
        {
          title: 'Correct',
          value: '8/10',
          icon: faChevronDown,
          iconColor: '#333',
        },
        { title: 'Fastest', value: '3.2s' },
      ],
    },
    {
      type: 'table-row',
      badge: 2,
      value: 'Bob Brown',
      progress: 50,
    },
  ]

  it('renders one row per item', () => {
    const { container } = render(<GameResultTable items={items} />)
    const rows = container.querySelectorAll('.tableRow')
    expect(rows.length).toBe(2)
    expect(container).toMatchSnapshot()
  })

  it('renders badge and value, and sets title attribute on value', () => {
    const { container } = render(<GameResultTable items={items} />)
    expect(screen.getByText('1')).toBeInTheDocument()
    expect(screen.getByText('2')).toBeInTheDocument()
    expect(screen.getByText('Alice Anderson')).toBeInTheDocument()
    expect(screen.getByText('Bob Brown')).toBeInTheDocument()

    const valueEls = Array.from(
      container.querySelectorAll('.tableRow .value'),
    ) as HTMLElement[]
    const aliceValueEl = valueEls.find(
      (el) => el.textContent === 'Alice Anderson',
    ) as HTMLElement
    const bobValueEl = valueEls.find(
      (el) => el.textContent === 'Bob Brown',
    ) as HTMLElement
    expect(aliceValueEl).toHaveAttribute('title', 'Alice Anderson')
    expect(bobValueEl).toHaveAttribute('title', 'Bob Brown')

    expect(container).toMatchSnapshot()
  })

  it('renders one chevron icon in the main area per row', () => {
    const { container } = render(<GameResultTable items={items} />)
    const rows = Array.from(container.querySelectorAll('.tableRow'))
    expect(rows.length).toBe(items.length)
    rows.forEach((row) => {
      const chevrons = row.querySelectorAll(
        '.main svg[data-icon="chevron-down"]',
      )
      expect(chevrons.length).toBe(1)
    })
  })

  it('hides details by default and toggles them on row click', () => {
    const { container } = render(<GameResultTable items={items} />)
    const rows = container.querySelectorAll('.tableRow')
    const firstRow = rows.item(0) as HTMLElement
    const firstRowDetails = firstRow.querySelector('.details') as HTMLElement

    expect(firstRowDetails.className).not.toContain('active')
    fireEvent.click(firstRow)

    expect(firstRowDetails.className).toContain('active')
    expect(screen.getByText('Correct')).toBeInTheDocument()
    expect(screen.getByText('8/10')).toBeInTheDocument()
    expect(screen.getByText('Fastest')).toBeInTheDocument()
    expect(screen.getByText('3.2s')).toBeInTheDocument()

    fireEvent.click(firstRow)
    expect(firstRowDetails.className).not.toContain('active')

    expect(container).toMatchSnapshot()
  })

  it('does not crash when details are missing', () => {
    const minimal: TableItem[] = [
      {
        type: 'table-row',
        badge: 3,
        value: 'Charlie',
        progress: 10,
      },
    ]
    const { container } = render(<GameResultTable items={minimal} />)
    const row = container.querySelector('.tableRow') as HTMLElement
    const details = row.querySelector('.details') as HTMLElement
    fireEvent.click(row)
    expect(details.className).toContain('active')

    expect(container).toMatchSnapshot()
  })

  it('renders separator correctly', () => {
    const { container } = render(
      <GameResultTable items={[{ type: 'table-separator' }]} />,
    )
    const rows = container.querySelectorAll('.tableSeparator')
    expect(rows.length).toBe(1)
    expect(container).toMatchSnapshot()
  })
})
