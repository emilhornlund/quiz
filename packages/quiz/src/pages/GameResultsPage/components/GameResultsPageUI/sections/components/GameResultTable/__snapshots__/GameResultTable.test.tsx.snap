// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GameResultTable > does not crash when details are missing 1`] = `
<div>
  <div
    class="table"
  >
    <div
      class="tableRow"
    >
      <div
        class="main"
      >
        <div
          class="badge"
        >
          3
        </div>
        <div
          class="value"
          title="Charlie"
        >
          Charlie
        </div>
        <div>
          <svg
            class="circularProgressBar correct small"
            height="60"
            width="60"
          >
            <circle
              class="circleBackground"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-width="6"
            />
            <circle
              class="circleProgress"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-linecap="round"
              stroke-width="6"
              style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 152.68140296446393;"
            />
            <text
              class="progressText"
              dominant-baseline="middle"
              text-anchor="middle"
              x="50%"
              y="50%"
            >
              10
              %
            </text>
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down"
            color="black"
            data-icon="chevron-down"
            data-prefix="fas"
            role="img"
            viewBox="0 0 448 512"
          >
            <path
              d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="details active"
      />
    </div>
  </div>
</div>
`;

exports[`GameResultTable > hides details by default and toggles them on row click 1`] = `
<div>
  <div
    class="table"
  >
    <div
      class="tableRow"
    >
      <div
        class="main"
      >
        <div
          class="badge"
        >
          1
        </div>
        <div
          class="value"
          title="Alice Anderson"
        >
          Alice Anderson
        </div>
        <div>
          <svg
            class="circularProgressBar correct small"
            height="60"
            width="60"
          >
            <circle
              class="circleBackground"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-width="6"
            />
            <circle
              class="circleProgress"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-linecap="round"
              stroke-width="6"
              style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 30.53628059289281;"
            />
            <text
              class="progressText"
              dominant-baseline="middle"
              text-anchor="middle"
              x="50%"
              y="50%"
            >
              82
              %
            </text>
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down"
            color="black"
            data-icon="chevron-down"
            data-prefix="fas"
            role="img"
            viewBox="0 0 448 512"
          >
            <path
              d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="details"
      >
        <div
          class="item"
        >
          <div
            class="title"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down"
              color="#333"
              data-icon="chevron-down"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
            Correct
          </div>
          <div
            class="value"
          >
            8/10
          </div>
        </div>
        <div
          class="item"
        >
          <div
            class="title"
          >
            Fastest
          </div>
          <div
            class="value"
          >
            3.2s
          </div>
        </div>
      </div>
    </div>
    <div
      class="tableRow"
    >
      <div
        class="main"
      >
        <div
          class="badge"
        >
          2
        </div>
        <div
          class="value"
          title="Bob Brown"
        >
          Bob Brown
        </div>
        <div>
          <svg
            class="circularProgressBar correct small"
            height="60"
            width="60"
          >
            <circle
              class="circleBackground"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-width="6"
            />
            <circle
              class="circleProgress"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-linecap="round"
              stroke-width="6"
              style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 84.82300164692441;"
            />
            <text
              class="progressText"
              dominant-baseline="middle"
              text-anchor="middle"
              x="50%"
              y="50%"
            >
              50
              %
            </text>
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down"
            color="black"
            data-icon="chevron-down"
            data-prefix="fas"
            role="img"
            viewBox="0 0 448 512"
          >
            <path
              d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="details"
      />
    </div>
  </div>
</div>
`;

exports[`GameResultTable > renders badge and value, and sets title attribute on value 1`] = `
<div>
  <div
    class="table"
  >
    <div
      class="tableRow"
    >
      <div
        class="main"
      >
        <div
          class="badge"
        >
          1
        </div>
        <div
          class="value"
          title="Alice Anderson"
        >
          Alice Anderson
        </div>
        <div>
          <svg
            class="circularProgressBar correct small"
            height="60"
            width="60"
          >
            <circle
              class="circleBackground"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-width="6"
            />
            <circle
              class="circleProgress"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-linecap="round"
              stroke-width="6"
              style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 30.53628059289281;"
            />
            <text
              class="progressText"
              dominant-baseline="middle"
              text-anchor="middle"
              x="50%"
              y="50%"
            >
              82
              %
            </text>
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down"
            color="black"
            data-icon="chevron-down"
            data-prefix="fas"
            role="img"
            viewBox="0 0 448 512"
          >
            <path
              d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="details"
      >
        <div
          class="item"
        >
          <div
            class="title"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down"
              color="#333"
              data-icon="chevron-down"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
            Correct
          </div>
          <div
            class="value"
          >
            8/10
          </div>
        </div>
        <div
          class="item"
        >
          <div
            class="title"
          >
            Fastest
          </div>
          <div
            class="value"
          >
            3.2s
          </div>
        </div>
      </div>
    </div>
    <div
      class="tableRow"
    >
      <div
        class="main"
      >
        <div
          class="badge"
        >
          2
        </div>
        <div
          class="value"
          title="Bob Brown"
        >
          Bob Brown
        </div>
        <div>
          <svg
            class="circularProgressBar correct small"
            height="60"
            width="60"
          >
            <circle
              class="circleBackground"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-width="6"
            />
            <circle
              class="circleProgress"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-linecap="round"
              stroke-width="6"
              style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 84.82300164692441;"
            />
            <text
              class="progressText"
              dominant-baseline="middle"
              text-anchor="middle"
              x="50%"
              y="50%"
            >
              50
              %
            </text>
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down"
            color="black"
            data-icon="chevron-down"
            data-prefix="fas"
            role="img"
            viewBox="0 0 448 512"
          >
            <path
              d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="details"
      />
    </div>
  </div>
</div>
`;

exports[`GameResultTable > renders one row per item 1`] = `
<div>
  <div
    class="table"
  >
    <div
      class="tableRow"
    >
      <div
        class="main"
      >
        <div
          class="badge"
        >
          1
        </div>
        <div
          class="value"
          title="Alice Anderson"
        >
          Alice Anderson
        </div>
        <div>
          <svg
            class="circularProgressBar correct small"
            height="60"
            width="60"
          >
            <circle
              class="circleBackground"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-width="6"
            />
            <circle
              class="circleProgress"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-linecap="round"
              stroke-width="6"
              style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 30.53628059289281;"
            />
            <text
              class="progressText"
              dominant-baseline="middle"
              text-anchor="middle"
              x="50%"
              y="50%"
            >
              82
              %
            </text>
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down"
            color="black"
            data-icon="chevron-down"
            data-prefix="fas"
            role="img"
            viewBox="0 0 448 512"
          >
            <path
              d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="details"
      >
        <div
          class="item"
        >
          <div
            class="title"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down"
              color="#333"
              data-icon="chevron-down"
              data-prefix="fas"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
            Correct
          </div>
          <div
            class="value"
          >
            8/10
          </div>
        </div>
        <div
          class="item"
        >
          <div
            class="title"
          >
            Fastest
          </div>
          <div
            class="value"
          >
            3.2s
          </div>
        </div>
      </div>
    </div>
    <div
      class="tableRow"
    >
      <div
        class="main"
      >
        <div
          class="badge"
        >
          2
        </div>
        <div
          class="value"
          title="Bob Brown"
        >
          Bob Brown
        </div>
        <div>
          <svg
            class="circularProgressBar correct small"
            height="60"
            width="60"
          >
            <circle
              class="circleBackground"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-width="6"
            />
            <circle
              class="circleProgress"
              cx="30"
              cy="30"
              fill="transparent"
              r="27"
              stroke-linecap="round"
              stroke-width="6"
              style="stroke-dasharray: 169.64600329384882; stroke-dashoffset: 84.82300164692441;"
            />
            <text
              class="progressText"
              dominant-baseline="middle"
              text-anchor="middle"
              x="50%"
              y="50%"
            >
              50
              %
            </text>
          </svg>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down"
            color="black"
            data-icon="chevron-down"
            data-prefix="fas"
            role="img"
            viewBox="0 0 448 512"
          >
            <path
              d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="details"
      />
    </div>
  </div>
</div>
`;
